File: Tests/toyMC_EffHistProd.py
from ROOT import (RooWorkspace, TH1F,
                  gSystem, RooDataHist,
                  TFile)
gSystem.Load('libP2VV')
from RooFitDecorators import *
import rootStyle

myStyle = rootStyle.plainstyle()
gROOT.SetStyle(myStyle.GetName())
gROOT.ForceStyle()
gStyle.UseCurrentStyle()

w = RooWorkspace("w")

#########################################
### Define variables and simple PDF's ###
#########################################
w.factory("t[0,10]")

#w.factory("TruthModel::tres(t)")
w.factory("GaussModel::tres(t,tres_m[0],tres_s[0.0005])")

w.factory("Decay::pdf(t,tau[2,0.01,4.0],tres,SingleSided)")
w.defineSet("observables","t")

##############################################
### Define acceptance function a la Wouter ###
##############################################
nbins = 100
interval = 10
effh1 = TH1F( "effh1", "effh1", nbins, w.var('t').getMin(), w.var('t').getMax()) 
for i in range(1, int(0.6 * nbins)):
    effh1.SetBinContent(i, 0.1 * i)
for i in range(int(0.6 * nbins), nbins + 1):
    effh1.SetBinContent(i, 1) 
effdatahist = RooDataHist("effdatahist", "effdatahist", RooArgList(w.var('t')), effh1) 

_import = getattr(w, 'import')
_import(effdatahist)

w.factory("HistFunc::eff(t, effdatahist)")
w.factory("EffHistProd::acc_pdf(pdf, eff)")
w.addClassDeclImportDir("..")
w.addClassImplImportDir("..")
w.importClassCode()

acc_pdf = w.pdf('acc_pdf')

from ROOT import RooFit, RooGenFitStudy
from ROOT import RooStudyManager

gfs = RooGenFitStudy();
gfs.storeDetailedOutput(True)
gfs.setGenConfig("acc_pdf", "t", RooFit.NumEvents(10000))
gfs.setFitConfig("acc_pdf", "t", RooFit.PrintLevel(-1), RooFit.Minimizer("Minuit2"))

mgr = RooStudyManager(w, gfs)

mgr.run(100)
## mgr.prepareBatchInput("aap", 100, kTRUE) ;

data = gfs.summaryData()
details = gfs.detailedData()
root_file = TFile.Open('data1D_%d.root' % nbins, 'recreate')
root_file.WriteTObject(data, 'data')
root_file.WriteTObject(details, 'details')
root_file.Close()

File: toys/toyMC_EffHistProd.cxx
#include <TH1.h>
#include <TSystem.h>
#include <TCanvas.h>

#include <RooGlobalFunc.h>
#include <RooWorkspace.h>
#include <RooDataHist.h>
#include <RooRealVar.h>
#include <RooStudyManager.h>
#include <RooGenFitStudy.h>
#include <RooDataSet.h>
#include <RooPullVar.h>
#include <RooPlot.h>
#include <RooPullVar.h>

using namespace RooFit;

RooWorkspace* toyMC() {
   gSystem->Load("libP2VV");
   RooWorkspace* w = new RooWorkspace("w");

   w->factory("t[0,10]");

   w->factory("GaussModel::tres(t,tres_m[0],tres_s[0.0005])");

   w->factory("Decay::pdf(t,tau[2,0.01,4.0],tres,SingleSided)");
   w->defineSet("observables","t");

   unsigned int nbins = 10;
   TH1F* effh1 = new TH1F( "effh1", "effh1", nbins, w->var("t")->getMin(), w->var("t")->getMax());
   for (unsigned int i = 1; i < 6; ++i) {
      effh1->SetBinContent(i, 0.1 * i);
   }
   for (unsigned int i = 6; i < nbins + 1; ++i) {
      effh1->SetBinContent(i, 1);
   }
   RooDataHist effdatahist("effdatahist", "effdatahist", RooArgList(*(w->var("t"))), effh1);

   w->import(effdatahist);

   w->factory("HistFunc::eff(t, effdatahist)");
   w->factory("EffHistProd::acc_pdf(pdf, eff)");
   w->addClassDeclImportDir("..");
   w->addClassImplImportDir("..");
   w->importClassCode();

   RooAbsPdf* acc_pdf = w->pdf("acc_pdf");

   RooDataSet* data = acc_pdf->generate(*(w->set("observables")), 10000);

   TCanvas* canvas = new TCanvas("canvas", "canvas", 500, 500);
   canvas->cd(1);
   RooPlot* frame = w->var("t")->frame(RooFit::Bins(30));
   data->plotOn(frame, RooFit::MarkerSize(0.5), RooFit::XErrorSize(0));
   acc_pdf->plotOn(frame, RooFit::LineWidth(2));
   frame->Draw();

   // RooGenFitStudy gfs;
   // gfs.setGenConfig("acc_pdf", "t", NumEvents(10000));
   // gfs.setFitConfig("acc_pdf", "t", PrintLevel(-1));

   // RooStudyManager mgr(*w, gfs);

   // mgr.run(1000);
   // mgr.prepareBatchInput("aap", 100, kTRUE);

   // RooDataSet* data = gfs.summaryData();
   // const RooArgSet* args = data->get();
   // RooPullVar tau_pull("tau_pull", "tau_pull", *(static_cast<RooRealVar*>(args->find("tau"))), *(w->var("t")));
   // data->addColumn(tau_pull);

   // TCanvas* canvas = new TCanvas("canvas", "canvas", 1000, 1000);
   // canvas->Divide(2, 2);
   // canvas->cd(1);

   return w;
}
